### PR TITLE
Introduction to Code Quality Metrics

### PR DESCRIPTION
**Pull request by wkurniawan07**

### CODE REVIEW

**Comment by wkurniawan07** [comment]
@kychua hope you don't mind reviewing!

File: contents/codeQuality/CodeQualityMetrics.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+> Since quantitative measurements are essential in all sciences, there is a continuous effort by computer science practitioners and theoreticians to bring similar approaches to software development.
+> The goal is obtaining objective, reproducible and quantifiable measurements, which may have numerous valuable applications in schedule and budget planning, cost estimation, quality assurance testing, software debugging, software performance optimization, and optimal personnel task assignments.
+>
+> https://en.wikibooks.org/wiki/Introduction_to_Software_Engineering/Quality/Metrics
+
+In fact, [the Boeing 777 project attributed its success to excellent usage of Quality Metrics (both code and non-code)](https://thesai.org/Downloads/Volume3No1/Paper%2021-Survey%20on%20Impact%20of%20Software%20Metrics%20on%20Software%20Quality.pdf).
+
+*Note that the term "Code Quality" here is scoped to be the readability, understandability, and maintainability of the code. While performance can be considered a legitimate basis for some metrics (e.g. time taken to run certain operations, based on profiling), it is not considered for this purpose.*
+
+## Example of Metrics
```
Comment (kychua) [inline contents/codeQuality/CodeQualityMetrics.md]: example**s**

File: contents/codeQuality/CodeQualityMetrics.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+### Complexity Metrics
+
+Complexity metrics measures how "complex" methods, classes, packages, etc. are. It is arguably the most useful metric for the largest number of developer because it is the easiest to grasp, the most directly relevant to the coding activity, and applicable to most/all programming languages.
```
Comment (kychua) [inline contents/codeQuality/CodeQualityMetrics.md]: Complexity metrics **measure** how "complex"...
...largest number of developer**s**...

Also, what does it mean by "complex"?

File: contents/codeQuality/CodeQualityMetrics.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+- The most well-known complexity metric is **Cyclomatic Complexity (CC)**, invented by Thomas McCabe in 1976. [Here](http://www.whiteboxtest.com/cyclomatic-complexity.php) is an excellent explanation on how to make use of it to measure your code quality. CC corresponds to the **minimum number of test cases needed to achieve 100% branch coverage**.
```
Comment (kychua) [inline contents/codeQuality/CodeQualityMetrics.md]: maybe swap 2nd and 3rd sentences for better flow - I'd want to know what CC is before knowing where I can learn more about it

File: contents/codeQuality/CodeQualityMetrics.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+- An alternative, less well-known, harder to calculate complexity metric is **NPath Complexity (NC)**, invented by Brian A. Nejmeh in 1988. NC corresponds to the **minimum number of test cases needed to achieve 100% path coverage**.
+
+> [Here](http://www.literateprogramming.com/mccabe.pdf) is the original paper by McCabe, and [here](http://delivery.acm.org/10.1145/50000/42379/p188-nejmeh.pdf?ip=137.132.250.8&id=42379&acc=ACTIVE%20SERVICE&key=FF6731C4D3E3CFFF%2EBB5EB8D2067C1662%2E4D4702B0C3E38B35%2E4D4702B0C3E38B35&CFID=726592962&CFTOKEN=31567842&__acm__=1486743928_30f8e35ed72667612c6449afe5db3998) is the original paper by Nejmeh.
```
Comment (kychua) [inline contents/codeQuality/CodeQualityMetrics.md]: link for Nejmeh paper doesn't seem to work

File: contents/codeQuality/CodeQualityMetrics.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+This complexity metrics can be extended to class level (e.g. summing or averaging the complexity values of all methods in a class), package level, or even project level.
```
Comment (kychua) [inline contents/codeQuality/CodeQualityMetrics.md]: **These** complexity metrics...

File: contents/codeQuality/CodeQualityMetrics.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+### Class Design Metrics
+
+Class design metrics measures how "well-designed" a class is. Metrics on this level are more applicable for QA team members and software architects, but still holds some relevance to students and junior developers who have to design a class-level API.
```
Comment (kychua) [inline contents/codeQuality/CodeQualityMetrics.md]: What does it mean by "well-designed"? How is this different from measuring complexity?

File: contents/codeQuality/CodeQualityMetrics.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+## Define Your Own Metric
+
+You would have noticed that these metrics are self-formulated rather than being inherent properties of a piece of software, and you are right.
+
+You might then ask: can I define my own code quality metric? Yes, you can. [This page](http://www.developer.com/tech/article.php/3644656/Software-Quality-Metrics.htm) has a good guide on how to formulate your own metric.
+
+## Making Sense of It
+
+Any code quality metric is as good as it is used; without context, it is merely a number. **Do not** use any metric just for the sake of it.
```
Comment (kychua) [inline contents/codeQuality/CodeQualityMetrics.md]: "Any code quality metric is as good as it is used" -> don't really understand this

**Review by kychua — CHANGES_REQUESTED** [review]
Good flow, informative & comprehensive.

For structure, can consider putting 'class design metrics' and 'package design metrics' under a 'Design Metrics' section so it'll be on the same level as 'Complexity Metrics'.

**Comment by wkurniawan07** [comment]
@kychua thanks for the review! Will address them by tonight.

File: contents/codeQuality/CodeQualityMetrics.md (hunk 2 above)
Comment (wkurniawan07) [reply contents/codeQuality/CodeQualityMetrics.md]: Will be fixed. Complex here means difficult to understand and to maintain, which I also have added.

File: contents/codeQuality/CodeQualityMetrics.md (hunk 4 above)
Comment (wkurniawan07) [reply contents/codeQuality/CodeQualityMetrics.md]: As it turns out, the link is generated only for short-term usage (as evident from the "ip" parameter in the web link). Will link to the [citation page](http://dl.acm.org/citation.cfm?doid=42372.42379) instead.

File: contents/codeQuality/CodeQualityMetrics.md (hunk 6 above)
Comment (wkurniawan07) [inline contents/codeQuality/CodeQualityMetrics.md]: Will explain :+1:

File: contents/codeQuality/CodeQualityMetrics.md (hunk 7 above)
Comment (wkurniawan07) [reply contents/codeQuality/CodeQualityMetrics.md]: Missed the word "how"; meant to say "as good as how it is used".

File: contents/codeQuality/CodeQualityMetrics.md (hunk 6 above)
Comment (wkurniawan07) [reply contents/codeQuality/CodeQualityMetrics.md]: In essence, we have learnt how "high cohesion, low coupling, promoting encapsulation, information hiding" are good SE principles. Some of the class design metrics can measure those (that's how powerful metrics are!) and well-designed classes conforms to those, which subsequently makes reuse, refactoring, and maintenance easier.

Complexity is a different concept as it determines just how much effort is needed to understand the class, although it has some effect to how well-designed a class is (e.g the WMC metric).

In any case, it's all explained in the chapter now.

File: contents/codeQuality/CodeQualityMetrics.md [hunk 8]
```diff
@@ -18,22 +18,26 @@ In fact, [the Boeing 777 project attributed its success to excellent usage of Qu
...
+> [Here](http://www.literateprogramming.com/mccabe.pdf) is the original paper by McCabe, and [here](http://dl.acm.org/citation.cfm?doid=42372.42379) is the original paper by Nejmeh.
 
-This complexity metrics can be extended to class level (e.g. summing or averaging the complexity values of all methods in a class), package level, or even project level.
+These complexity metrics can be extended to class level (e.g. summing or averaging the complexity values of all methods in a class), package level, or even project level.
 
 ### Class Design Metrics
 
-Class design metrics measures how "well-designed" a class is. Metrics on this level are more applicable for QA team members and software architects, but still holds some relevance to students and junior developers who have to design a class-level API.
+Class design metrics measures how "well-designed" a class is. "Well-designed" here is defined as conformance to good software engineering principles such as high cohesion, low coupling, promoting encapsulation and information hiding. Well-designed classes promotes reuse and eases maintenance effort.
```
Comment (kychua) [inline contents/codeQuality/CodeQualityMetrics.md]: Class design metrics **measure**... 
Well-designed classes **promote** and *reduce* maintenance effort

File: contents/codeQuality/CodeQualityMetrics.md [hunk 9]
```diff
@@ -18,22 +18,26 @@ In fact, [the Boeing 777 project attributed its success to excellent usage of Qu
...
+
+Metrics on this level are more applicable for QA team members and software architects, but still holds some relevance to students and junior developers who have to design a class-level API.
```
Comment (kychua) [inline contents/codeQuality/CodeQualityMetrics.md]: ...but still **hold** some relevance...

**Review by kychua — CHANGES_REQUESTED** [review]

**Review by kychua — APPROVED** [review]

**Comment by kychua** [comment]
sorry for the late review, thought I'd accepted this already!

File: contents/codeQuality/CodeQualityMetrics.md [hunk 10]
```diff
@@ -0,0 +1,102 @@
+# Code Quality Metrics
```
Comment (damithc) [inline contents/codeQuality/CodeQualityMetrics.md:1]: The article seems to assume an OOP code? If so, mention that upfront?

File: contents/codeQuality/CodeQualityMetrics.md [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+Any code quality metric is as good as how it is used; without context, it is merely a number. **Do not** use any metric just for the sake of it.
+- It makes little sense to say: "This method has a CC value of 42. It is bad."
+- It makes more sense to say: "This method has a CC value of 42. That means we need at least 42 test cases to achieve full branch coverage for it, and it is bad."
+- It makes the most sense to say: "This method has a CC value of 42. That means we need at least 42 test cases to achieve full branch coverage for it. This indicates an overly complex method, and it will be difficult to maintain in the long run. Try to separate it to calls of a few methods of CC value of 10 or less each; that way we can better design test cases for those smaller methods and achieve perfect coverage."
+
+[Here](http://homepages.dcc.ufmg.br/~figueiredo/disciplinas/lectures/detection-strategy-examples_v01.pdf) is an excellent example of how different metrics are used to determine which classes/methods need some maintenance works.
+
+## Code Quality Metrics Tools
+
```
Comment (damithc) [inline contents/codeQuality/CodeQualityMetrics.md:90]: Perhaps mentions these as 'examples'

File: contents/codeQuality/CodeQualityMetrics.md [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+> Since quantitative measurements are essential in all sciences, there is a continuous effort by computer science practitioners and theoreticians to bring similar approaches to software development.
+> The goal is obtaining objective, reproducible and quantifiable measurements, which may have numerous valuable applications in schedule and budget planning, cost estimation, quality assurance testing, software debugging, software performance optimization, and optimal personnel task assignments.
+>
+> https://en.wikibooks.org/wiki/Introduction_to_Software_Engineering/Quality/Metrics
+
+In fact, [the Boeing 777 project attributed its success to excellent usage of Quality Metrics (both code and non-code)](https://thesai.org/Downloads/Volume3No1/Paper%2021-Survey%20on%20Impact%20of%20Software%20Metrics%20on%20Software%20Quality.pdf).
+
+*Note that the term "Code Quality" here is scoped to be the readability, understandability, and maintainability of the code. While performance can be considered a legitimate basis for some metrics (e.g. time taken to run certain operations, based on profiling), it is not considered for this purpose.*
```
Comment (damithc) [inline contents/codeQuality/CodeQualityMetrics.md:19]: We usually use single quotes e.g. 'Code Quality' unless it is something said by a person.

**Review by damithc — COMMENTED** [review]
Good work @wkurniawan07 
Some comments added. Revise as you see fit.
Adding more content and/or getting more peer reviews is welcome but not required. 
I will be merging end of week 12

**Pull request merged**